Rubberduck.VBEEditor/Class1.cs

```diff
@@ -0,0 +1,964 @@
+            Marshal.AddRef(_outerObject);
+        }
+
+        private bool _isDisposed;
+        public void Dispose()
+        {
+            if (_isDisposed) return;
+            _isDisposed = true;
+            Marshal.Release(_outerObject);
+        }
+
+        public CustomQueryInterfaceResult GetInterface(ref Guid iid, out IntPtr ppv)
+        {
+            ppv = IntPtr.Zero;
+            if ((!_isDisposed) && (iid == typeof(T).GUID))       // no need to offer IID_IUnknown here, as it is handled by the aggregation object
+            {
+                ppv = _outerObject;
+                Marshal.AddRef(_outerObject);
+                return CustomQueryInterfaceResult.Handled;
+            }
+            return CustomQueryInterfaceResult.Failed;
+        }
+    }
+
+    [ComImport(), Guid("00020400-0000-0000-C000-000000000046")]
+    [InterfaceType(ComInterfaceType.InterfaceIsIDispatch)]
+    public interface IDispatch
+    {
+    }
+
+    // A compatible version of ITypeInfo, where COM objects are outputted as IntPtrs instead of objects
+    [ComImport(), Guid("00020401-0000-0000-C000-000000000046")]
+    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+    public interface ITypeInfo_VBE
+    {
+        void GetTypeAttr(out IntPtr ppTypeAttr);
+        void GetTypeComp(out IntPtr ppTComp);
+        void GetFuncDesc(int index, out IntPtr ppFuncDesc);
+        void GetVarDesc(int index, out IntPtr ppVarDesc);
+        void GetNames(int memid, [Out] out string rgBstrNames, int cMaxNames, out int pcNames);
+        void GetRefTypeOfImplType(int index, out int href);
+        void GetImplTypeFlags(int index, out IMPLTYPEFLAGS pImplTypeFlags);
+        void GetIDsOfNames(string[] rgszNames, int cNames, int[] pMemId);
+        void Invoke(object pvInstance, int memid, short wFlags, ref DISPPARAMS pDispParams, IntPtr pVarResult, IntPtr pExcepInfo, out int puArgErr);
+        void GetDocumentation(int index, out string strName, out string strDocString, out int dwHelpContext, out string strHelpFile);
+        void GetDllEntry(int memid, INVOKEKIND invKind, IntPtr pBstrDllName, IntPtr pBstrName, IntPtr pwOrdinal);
+        void GetRefTypeInfo(int hRef, out IntPtr ppTI);
+        void AddressOfMember(int memid, INVOKEKIND invKind, out IntPtr ppv);
+        void CreateInstance(object pUnkOuter, ref Guid riid, out object ppvObj);
+        void GetMops(int memid, out string pBstrMops);
+        void GetContainingTypeLib(out IntPtr ppTLB, out int pIndex);
+        void ReleaseTypeAttr(IntPtr pTypeAttr);
+        void ReleaseFuncDesc(IntPtr pFuncDesc);
+        void ReleaseVarDesc(IntPtr pVarDesc);
+
+        void Placeholder1();
+        IDispatch GetStdModInstance();            // a handy extra vtable entry we can use to invoke members in standard modules.
+    }
+
+    // FIXME there's probably some better builtin c# class for this
+    public class DisposableList<T> : List<T>, IDisposable
+        where T : class
```

Comment:
you want to restrict `T` to `IDisposable` otherwise the cast below isn't guaranteed to work. 

```
public class DisposableList<T> : List<T>, IDisposable 
    where T : IDisposable
{
```

Reply:
Usually I'd admonish composition over inheritance here, but this seems like a pretty good use of inheritance.
It might be more in order to use `IList` though:

```
public class DisposableList<T> : IList<T>, IDisposable
    where T : IDisposable
{
    private readonly IList<T> wrappedList;
    
    public DisposableList<T>(IList<T> wrappedList) {
         this.wrappedList = wrappedList;
    }
    // ...
```

The problem with that is: we don't have any control over outside references of our IList...